src/navier_stokes/StaggeredStokesProjectionPreconditioner.cpp

```diff
@@ -215,10 +218,20 @@ StaggeredStokesProjectionPreconditioner::solveSystem(SAMRAIVectorReal<NDIM, doub
     // U^* := inv(rho/dt - K*mu*L) F_U
     //
     // An approximate Helmholtz solver is used.
-    d_velocity_solver->setHomogeneousBc(true);
-    auto p_velocity_solver = dynamic_cast<LinearSolver*>(d_velocity_solver.getPointer());
-    if (p_velocity_solver) p_velocity_solver->setInitialGuessNonzero(false);
-    d_velocity_solver->solveSystem(*U_vec, *F_U_vec);
+    if (d_needs_velocity_solver)
+    {
+        d_velocity_solver->setHomogeneousBc(true);
+        auto p_velocity_solver = dynamic_cast<LinearSolver*>(d_velocity_solver.getPointer());
+        if (p_velocity_solver) p_velocity_solver->setInitialGuessNonzero(false);
+        d_velocity_solver->solveSystem(*U_vec, *F_U_vec);
+    }
+    else
+    {
+        TBOX_ASSERT(!d_U_problem_coefs.cIsZero());
+        d_velocity_data_ops->scale(U_vec->getComponentDescriptorIndex(0),
+                                   1.0 / d_U_problem_coefs.getCConstant(),
+                                   F_U_vec->getComponentDescriptorIndex(0));
+    }
 
     // (2) Solve the pressure sub-problem.
```

Comment:
@boyceg : are you using a diagonal preconditioning here?

Reply:
Not really, this is just making sure that the the units are correct for evaluating the RHS in the next step.

Reply:
It would be very easy to turn this into a proper diagonal preconditioner in the constant coefficient case by adding an additional term here.

Reply:
It would not be too hard to turn this into a diagonal preconditioner, but it would require looking up the grid spacings etc.